adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -46,7 +48,190 @@ def _normsqr_groups(grads):
     return np.array(ret)
 
 
-class AdaScale(object):
+class _LocalVariables(collections.abc.MutableMapping):
+    """
+    A dictionary that remembers initial values and has a ``reset`` function
+    to reset data to its initial state.
+    """
+    def __init__(self):
+        self.data_copy = {}
+        self.data = {}
+
+    def set_initial_values(self, data: dict):
+        self.data_copy = data
+        self.reset()
+
+    def reset(self):
+        self.data = copy.deepcopy(self.data_copy)
+
+    def __getitem__(self, item):
+        return self.data.__getitem__(item)
+
+    def __setitem__(self, key, value):
+        self.data.__setitem__(key, value)
+
+    def __delitem__(self, key):
+        self.data.__delitem__(key)
+
+    def __iter__(self):
+        return self.data.__iter__()
+
+    def __len__(self):
+        return self.data.__len__()
+
+
+class ScalingRuleBase(object):
+    """
+    Base class for scaling rules.
+
+    Arguments:
+        optimizer (torch.optim.Optimizer): Optimizer to apply this scaling rule
+        to.
+        scale (float): Scaling factor of the batch size, e.g. using a 10x
+            larger batch size (summed across all replicas) means a scale of
+            10. If None, defaults to ``num_replicas``.
+        num_replicas (int): Number of replicas for distributed training. If
+            None, defaults to ``torch.distributed.get_world_size()``.
+        patch_optimizer (bool): If True, monkey-patches the ``step`` method of
+            the optimizer with this scaling rule's ``step`` method.
+    """
```

Comment:
The overall logic is split between classes in a way that's not quite ideal currently, and probably makes the lr scaling class a bit more complicated than it needs to be. In particular, the norm and var estimation logic and get_progress logic should be outside of the scaling rule, because they should be the same regardless of the scaling rule, and only certain scaling rules need to access them (ie. adascale).

Reply:
I'm assuming that by "should be outside of the scaling rule" you mean that they need to be moved to the base class.
I understand that norm and var estimation logic remain the same regardless of the scaling rule, but before we move the logic to the base class I want to make sure that they make sense to be attributes attached to the base class --- if we simply add all invariant mathematical calculations to the base class it will end up being a huge Helper/Utility Class, in which case I think these logics might make more sense to be hosted somewhere else.

Reply:
Maybe a more maintainable/unit-testable way is to create a separate `GradientNoiseScale` class that calculates the norm/var statistics, and add it as a field of the AdaptiveDataParallel class? I don't think it should go into the base scaling rule class either, because it's not really a property of the scaling rule.

Reply:
Tried to do that but it seems hard to completely decouple GradientNoiseScale from AdaScale. They are dependent on each other (GradientNoiseScale requires gradient accumulations and AdaScale requires stats calculations). So I made GradientNoiseScale a mixin class to add stats calculations to the ScalingRuleBase class.

Reply:
The scaling rule class is still more complicated than it needs to be. Essentially, the new `AdaScale` class should just (roughly) be the logic in the `step` method of the old AdaScale class. Everything else can be factored out.